src/persist-client/src/internal/machine.rs

```diff
@@ -343,21 +364,47 @@ where
         .await
     }
 
-    pub async fn heartbeat_reader(
+    pub async fn compare_and_downgrade_since<O: Codec64>(
         &mut self,
-        reader_id: &ReaderId,
+        reader_id: &CriticalReaderId,
+        expected_opaque: &O,
+        (new_opaque, new_since): (&O, &Antichain<T>),
+    ) -> (Result<Since<T>, (O, Since<T>)>, RoutineMaintenance) {
+        let metrics = Arc::clone(&self.metrics);
+        let (_seqno, res, maintenance) = self
+            .apply_unbatched_idempotent_cmd(
+                &metrics.cmds.compare_and_downgrade_since,
+                |_seqno, state| {
+                    state.compare_and_downgrade_since(
+                        reader_id,
+                        Opaque(Codec64::encode(expected_opaque)),
```

Comment:
IMO this is an odd place for the Codec64 encode/decode. I'd expect either it be entirely above Machine or down on StateCollections::compare_and_downgrade_since. preference for the latter so that we can store the `codec_name` on CriticalReaderState and do some verification that it matches